File: sdk-java/src/test/java/io/superdurable/dex/iwfcompat/AnyCommandCombinationWorkflow.java
/*
 * Portions of this file are derived from indeedeng/iwf-java-sdk.
 * Those portions are licensed under the Apache License, Version 2.0.
 * See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
 *
 * Modifications Copyright (c) 2026 Super Durable, Inc.
 *
 * Modifications are licensed under the Super Durable Source License 1.0.
 * Third-Party Materials remain under the Apache License, Version 2.0.
 * See LICENSE and LEGACY_NOTICES.md.
 */

package io.superdurable.dex.iwfcompat;

import io.superdurable.dex.Channel;
import io.superdurable.dex.Context;
import io.superdurable.dex.Flow;
import io.superdurable.dex.PersistenceSchema;
import io.superdurable.dex.RetryPolicy;
import io.superdurable.dex.Step;
import io.superdurable.dex.StepList;
import io.superdurable.dex.StepDecision;
import io.superdurable.dex.StepOptions;
import io.superdurable.dex.Wait;

final class AnyCommandCombinationWorkflow implements Flow<Integer> {
    private final Channel<Integer> first = Channel.define("test-signal-1", Integer.class);
    private final Channel<Integer> second = Channel.define("test-signal-2", Integer.class);
    private final Channel<Integer> third = Channel.define("test-signal-3", Integer.class);
    private final AnyCommandCombinationStep start =
            new AnyCommandCombinationStep(first, second, third);

    @Override
    public StepList<Integer> getSteps() {
        return StepList.startStep(start);
    }

    @Override
    public PersistenceSchema getPersistenceSchema() {
        return PersistenceSchema.of(first, second, third);
    }
}

final class AnyCommandCombinationStep implements Step<Integer> {
    private final Channel<Integer> first;
    private final Channel<Integer> second;
    private final Channel<Integer> third;

    AnyCommandCombinationStep(
            final Channel<Integer> first,
            final Channel<Integer> second,
            final Channel<Integer> third) {
        this.first = first;
        this.second = second;
        this.third = third;
    }

    @Override
    public Class<Integer> getInputType() {
        return Integer.class;
    }

    @Override
    public Wait waitFor(final Context context, final Integer input) {
        throw new IllegalArgumentException(
                "Found unknown condition ID in the combination list");
    }

    @Override
    public StepDecision execute(final Context context, final Integer input) {
        return StepDecision.gracefulComplete(input);
    }

    @Override
    public StepOptions getStepOptions() {
        return StepOptions.newBuilder()
                .waitForRetry(RetryPolicy.newBuilder().maximumAttempts(1).build())
                .build();
    }
}

File: sdk-rust/crates/dex-sdk/tests/integ/any_command_combination_workflow.rs
// Portions of this file are derived from indeedeng/iwf-java-sdk.
// Those portions are licensed under the Apache License, Version 2.0.
// See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
//
// Modifications Copyright (c) 2026 Super Durable, Inc.
//
// Modifications are licensed under the Super Durable Source License 1.0.
// Third-Party Materials remain under the Apache License, Version 2.0.
// See LICENSE and LEGACY_NOTICES.md.

use std::sync::LazyLock;

use dex_sdk::{
    Channel, Context, Flow, HandlerError, HandlerResult, PersistenceSchema, RetryPolicy, Step,
    StepDecision, StepList, StepOptions, Wait,
};

static FIRST: LazyLock<Channel<i32>> = LazyLock::new(|| Channel::new("test-signal-1"));
static SECOND: LazyLock<Channel<i32>> = LazyLock::new(|| Channel::new("test-signal-2"));
static THIRD: LazyLock<Channel<i32>> = LazyLock::new(|| Channel::new("test-signal-3"));

pub(crate) struct AnyCommandCombinationWorkflow {
    first: Channel<i32>,
    second: Channel<i32>,
    third: Channel<i32>,
    start: AnyCommandCombinationStep,
}

impl AnyCommandCombinationWorkflow {
    pub(crate) fn new() -> Self {
        Self {
            first: FIRST.clone(),
            second: SECOND.clone(),
            third: THIRD.clone(),
            start: AnyCommandCombinationStep,
        }
    }
}

impl Flow for AnyCommandCombinationWorkflow {
    type StartInput = i32;

    fn steps(&self) -> StepList<'_, Self::StartInput> {
        StepList::start(&self.start)
    }

    fn persistence(&self) -> PersistenceSchema {
        PersistenceSchema::new()
            .channel(&self.first)
            .channel(&self.second)
            .channel(&self.third)
    }
}

struct AnyCommandCombinationStep;

impl Step for AnyCommandCombinationStep {
    type Input = i32;

    fn wait_for(&self, _context: &mut Context, _input: i32) -> HandlerResult<Wait> {
        Err(HandlerError::new(
            "UnknownConditionId",
            "Found unknown condition ID in the combination list",
        ))
    }

    fn execute(&self, _context: &mut Context, input: i32) -> HandlerResult<StepDecision> {
        Ok(StepDecision::graceful_complete(input))
    }

    fn options(&self) -> StepOptions<Self::Input> {
        StepOptions::new().wait_for_retry(RetryPolicy::new().maximum_attempts(1))
    }
}
